[RFC] System containers

giuseppe opened the pull request:
these patches are the current status of system containers.  There are a few things missing on Atomic Host before this feature can be used:

- skopeo
- runc
- python3-gobject-base

For an idea of how they work, I prepared a demo here: https://asciinema.org/a/40300

runcom commented:
Awesome :tada:

giuseppe commented:
bot, retest this please

rhatdan commented on atomic:
```diff
@@ -273,6 +280,9 @@ if __name__ == '__main__':
+    imagesp.add_argument("--oci", dest="oci",
+                         action='store_true', default=False,
+                         help=_('Show OCI containers'))
 
```
Wouldn't the default show all container images?  Then we can add a selection criteria to limit the type of images.

--filter oci,docker,rkt?

What is an OCI IMage? Seems a lot like a docker image?

giuseppe replied:
I like this idea, just the output has a different format (as most of the Docker fields have no equivalent in an OSTree repository).  Is there anything depending on the format of the `atomic images` output?  How will we print both formats?

This is the output I have right now:

```
bash-4.3# atomic images
REPOSITORY                          TAG    IMAGE ID     CREATED             VIRTUAL SIZE
 docker.io/gscrivano/spc-helloworld latest sha256:aa510 2016-03-22 12:23    216.41 MB 
 docker.io/gscrivano/spc-flannel    latest sha256:9b5dd 2016-03-21 13:44    230.42 MB 
 docker.io/busybox                  latest sha256:47bcc 2016-03-18 18:22    1.11 MB   
 docker.io/alpine                   latest sha256:70c55 2016-03-02 17:16    4.79 MB   
bash-4.3# atomic images --oci
IMAGE                                    COMMIT                                                          
dockerimg/busybox-latest                 ff6b0e9239c9569be3ef48699519d7842975a48c98b0ba560000363006364c5e
dockerimg/gscrivano/spc-flannel-latest   00b316952aee741a2eeb62c98e7dc7a2018c6ce94b1e3cd93ba0ce12955c3cf3
```

rhatdan replied:
I like this, but I would want path to rootfs for --oci option.

giuseppe replied:
there is no rootfs assigned yet for the `--oci` flag.  These are the images available in the OSTree repository, which could have or not been checked-out.  The checkout from the OSTree repository is done by `install --oci`.  There could be different containers using the same image.

rhatdan commented on atomic:
```diff
@@ -327,6 +337,18 @@ if __name__ == '__main__':
+                       help=_("import from an OSTree repository."))
+    pullp.add_argument("image", help=_("image id"))
+    pullp.set_defaults(func='pull_image')
+
```
What is the default?  Is the default something we can set in /etc/atomic.conf?

giuseppe replied:
I am still undecided here about the format.  Should we have --tar/--docker/--ostree or we should rather use a prefix (as we do for install --oci) like:
```
atomic pull tar:///foo/bar.tar
atomic pull ostree://foo:bar
```

runcom replied:
I'm more for the image prefix as a way to standardize those flags

giuseppe replied:
You convinced me.  I added a patch that uses the image prefixes

rhatdan commented on Atomic/atomic.py:
```diff
@@ -254,6 +270,80 @@ def push(self):
+        enc = sys.getdefaultencoding()
+        for k, v in layers.items():
+            out = subprocess.check_output([ATOMIC_LIBEXEC + '/dockertar-sha256-helper',
+           v], stderr=DEVNULL)
```
Should this be overridden via ENV variable or search for it locally as well as in /usr/libexec/atomic, so that we can run tests before installing the package.

giuseppe replied:
ATOMIC_LIBEXEC is already constructed looking at the environment variable `ATOMIC_LIBEXEC` as:

```
ATOMIC_LIBEXEC = os.environ.get('ATOMIC_LIBEXEC', '/usr/libexec/atomic')
```
It is the same mechanism used for the `migrate.sh` script.  Is it fine like this?

rhatdan replied:
Ok I missed that.

rhatdan commented on Atomic/atomic.py:
```diff
@@ -254,6 +270,80 @@ def push(self):
+                        self._pull_dockertar_layers(repo, imagebranch, temp_dir, input_layers)
+            finally:
+                shutil.rmtree(temp_dir)
+        return
```
Should be in a separate method or probably in a separate PY file.

rhatdan commented on Atomic/atomic.py:
```diff
@@ -771,6 +881,343 @@ def install(self):
+
+    def _uninstall_oci_container(self, name):
+        ocidir = os.path.realpath("/var/lib/containers/atomic/%s" % name)
+        self.args.display = False
```
Should this be a constant defined above?

giuseppe replied:
yes, will do.

rhatdan commented on Atomic/atomic.py:
```diff
@@ -771,6 +881,343 @@ def install(self):
+        refs = {}
+        app_refs = []
+        prefix = "dockerimg/"
+
```
Where does this prefix come from?

giuseppe replied:
it is the prefix for OSTree branches generated from a Docker import.  I will define it in a constant to make it clearer

rhatdan commented:
Man Pages?
Bash Completion?

giuseppe commented:
I have not added man pages and bash completion yet.  I will do this as soon as we agree on the interface

rhatdan commented:
@runcom @mrunalp @baude @cgwalters PTAL

rhatdan commented:
Should the option be --oci or --ostree?  --oci, images should be able to be used in docker or on --ostree.  I believe I should be able to pull an oci image and then install it into docker or ontop of ostree.

giuseppe commented:
I am really confused on how it should be named.  I thought for a while that `--system` could be the most appropriate name, but in reality this mechanism can be used for any container, not only system containers, while `--oci` is just one of the possible ways that we can use this mechanism for.  Probably `--ostree` is the best option we have.

The format of the image, no matter how it is fetched (via Docker, Docker tarball or OSTree), is made of the rootfs+files under /exports (at the moment: `service.template`, `config.json.template`, `runtime.json.template`) that are distributed within the rootfs itself.  The only part which is related to OCI is that the files `config.json.template` and `runtime.json.template` are automatically copied from '/foo/container/rootfs/exports' to '/foo/container' (after some substitutions are made for values specified with `--set`).  It is assumed though that systemd is present; the entire life cycle of these containers is managed through systemctl.

rhatdan commented:
@giuseppe I think we need to have a phone call discussion on this CLI.  Lets bring in @runcom and @cgwalters

mrunalp commented on dockertar-sha256-helper.go:
```diff
@@ -0,0 +1,32 @@
+	sha_256 := sha256.New()
+	w := gzip.NewWriter(sha_256)
+
+	for true {
```
This could be just 
```
for {
...
}
```

mrunalp commented on dockertar-sha256-helper.go:
```diff
@@ -0,0 +1,32 @@
+		}
+		w.Write(buf[:n])
+	}
+	w.Close()
```
This could be moved up as a defer.

runcom commented on Atomic/atomic.py:
```diff
@@ -254,6 +273,54 @@ def push(self):
+        layers_map = {}
+        enc = sys.getdefaultencoding()
+        for k, v in layers.items():
+            out = subprocess.check_output([ATOMIC_LIBEXEC + '/dockertar-sha256-helper',
```
can this helper be just plain python instead of embedding another bin?

runcom replied:
or probably do we need gzip from golang?

mrunalp replied:
:+1: to keeping to to python if we can.

giuseppe replied:
it is ugly but we need gzip from golang :(

rhatdan replied:
Why?

giuseppe replied:
so that given a layer in a docker save'd tarball we can compute the same checksum it has on the docker registry (the checksum found in the registry manifest is computed after the .tar file is gzipped).  Other gzip implementations generate a different file, even though the original .tar is the same.

It is the issue discussed here: https://github.com/projectatomic/skopeo/issues/11

mrunalp commented on dockertar-sha256-helper.go:
```diff
@@ -0,0 +1,32 @@
+
+	for true {
+		n, err := reader.Read(buf)
+		if err != nil && err != io.EOF {
```
nit: Could process the `n > 0` case first as it as suggested in the documentation.
Something like
```
if n > 0 {
  w.Write(buf[:n])
}
if err == io.EOF {
  break
}
if err != nil {
  panic(err)
}

mrunalp commented on dockertar-sha256-helper.go:
```diff
@@ -0,0 +1,32 @@
+package main
```
nit: You could add an empty line after `package main`

mrunalp commented on Atomic/atomic.py:
```diff
@@ -771,6 +858,371 @@ def install(self):
+                    _write_template(infile.read(), values, outfile)
+
+        unitfile = os.path.join(exports, "service.template")
+        unitfileout = os.path.join("/etc/systemd/system", "%s.service" % name)
```
We could probably store `/etc/systemd/system` as a constant as it is being reused.

mrunalp commented on dockertar-sha256-helper.go:
```diff
@@ -0,0 +1,32 @@
+		if n <= 0 {
+			break
+		}
+		w.Write(buf[:n])
```
Should check the return code here.

mrunalp replied:
disregard the error check on Write comment as you're not writing to a file.

giuseppe replied:
thanks, I have addressed these in some followup commits

giuseppe commented:
I went through a major refactoring of the patches, that made impossible to use fixup commits.  I have pushed the updated patches here :arrow_up:

rhatdan commented:
I would like to start moving this to getting it merged, since we want to ship a new version of atomic

atomic-1.10 should include fixed atomic scan, this patch, docker-latest support, Atomic handling of $PWD, atomic unlock, And a bunch of other fixes.

rhatdan commented on Atomic/atomic.py:
```diff
@@ -30,6 +41,22 @@
+ATOMIC_LIBEXEC = os.environ.get('ATOMIC_LIBEXEC', '/usr/libexec/atomic')
+
+OSTREE_DOCKERIMG_PREFIX = "dockerimg/"
+SYSTEMD_UNIT_FILES_DEST = "/etc/systemd/system"
```
Why is this dockerimage, Could it be ociimage?

giuseppe replied:
this is really a Docker layer/image, and it is the prefix of the branch which points to the docker image layer, these OSTree branches will look like:
dockerimg/CHECKSUM (for each layer)
dockerimg/image-name (for the image, which has information on what layers are used)

rhatdan replied:
Is this hard coded into ostree?  If I explode oci images onto ostree will it do it in dockerimage?  Just want to minimize the use of docker if possible.

giuseppe replied:
no, this is only into `atomic`.  I kept two different namespaces for images that are imported directly from Docker (either registry or tarball) and those that are created directly into an OSTree commit.
In the first case we want to track every layer used to compose the image, so that it can be reused.  Each layer is mapped to an OSTree branch with the dockerimg/ prefix.

For images that are created directly as OSTree commits (as `ostree commit -b IMAGE_NAME rootfs`) there is no mapping with a Docker layer so there is no `dockerimg/` prefix.

We can "flatten" a Docker imported image by having a commit of the final rootfs, and this can be done at any stage, but doing so will lose all the advantages of tracking the Docker layers.

giuseppe replied:
thinking more of it, it should not make a difference if we call it in another way, e.g. `ociimage/`, being a Docker layer is just a detail of how we will handle it, so I am going to change it

rhatdan commented on Atomic/atomic.py:
```diff
@@ -735,15 +735,10 @@ def _uninstall_system_container(self):
+        shutil.rmtree("/var/lib/containers/atomic/%s.0" % self.name)
+        if os.path.exists("/var/lib/containers/atomic/%s.1" % self.name):
+            shutil.rmtree("/var/lib/containers/atomic/%s.1" % self.name)
 
```
Can we make this a constant  

CONTAINER_LIB=/var/lib/containers/atomic

runcom commented:
@giuseppe is this still blocked by me on adding `--insecure` in `skopeo`?

rhatdan commented on Atomic/atomic.py:
```diff
@@ -690,6 +803,405 @@ def install(self):
+        repo = self._get_ostree_repo()
+
+        revs = [x for x in repo.list_refs()[1] if x.startswith(OSTREE_DOCKERIMG_PREFIX) and len(x) != len(OSTREE_DOCKERIMG_PREFIX) + 64]
+        max_column = max([len(rev) for rev in revs]) + 2
```
DOCKERIMG or OCIIMAGE?

giuseppe replied:
I'll change it to OCIIMAGE

rhatdan commented on atomic:
```diff
@@ -501,6 +518,23 @@ if __name__ == '__main__':
+    updatecontainerp = subparser.add_parser(
+        "update-container", help=_("update a container"),
+        epilog="update a container.")
+    updatecontainerp.set_defaults(func='update_container')
```
How is this different then update? Can we just improve atomic update rather then add this command?

rhatdan commented on docs/atomic-install.1.md:
```diff
@@ -51,6 +53,12 @@ If --display is not specified the install command will execute.
+
+**--system**
+  Install a system container.  System containers should be used for services that are required early by the system or that cannot run in a normal container.
+
```
A system container is a container that is executed out of an systemd unit file early in boot, using ocitools and runc.   No need to define this as running early in boot.

giuseppe replied:
thanks, I will change it to "A system container is a container that is executed out of an systemd unit file early in boot, using runc.", as at the moment we are not using ocitools

rhatdan commented on docs/atomic-uninstall.1.md:
```diff
@@ -49,5 +50,8 @@ Any additional arguments will be appended to the command.
    If name is specified `atomic uninstall` will uninstall the named container from the system, otherwise it will uninstall the container images.
 
+**--system**
+   Uninstall a system container.
```
Is there anyways we can figure out that this is a system container?  Should we leave a bit of information around to tell us that the container is a system container so atomic uninstall will just do the right thing.  If I am running a system container and accidently do not include the --system flag, what happens?

giuseppe replied:
I thought these are two different namespaces, in theory it is possible to have a Docker container and a system container with the same name.  Should we not allow to install the system container if there is already a docker container with that name?  Though the user can still create the Docker container in another way after a system container with the same name is running.  How then it will be possible to remove only the system container with `atomic` (or viceversa, the Docker container)?

rhatdan replied:
I think we should do our best to prevent duplicate named containers.  As  you state we could end up with bad results when we use alternate names.  

If I uninstall a --system container, I would end up with a docker container which I could later uninstall.

If we prevent the installing of a --system container if a docker container exists, then we don't end up in this state unless a user went behind our back.  I think this is better then figuring out what happens if a user tries to remove a system container without specifying --system.

giuseppe commented:
@runcom I added `--insecure` to the skopeo invocation when :http: is used but I am not sure if we have to consider tihs like a blocking issue.  Maybe I can just remove this part for now?  It is just this line: https://github.com/projectatomic/atomic/pull/334/commits/44fe083e1ec36263a1129ab5e46a5bd97cfcd878#diff-555a8731c137f41d8ec2a5b76216bdecR872

rhatdan commented on Atomic/atomic.py:
```diff
@@ -272,6 +360,11 @@ def set_args(self, args):
         try:
+            self.system = args.system
+        except:
+            self.system = False
```
This is not needed since this is defaulted above, Shouldn't we just pass as we have on other variable assignment.

giuseppe replied:
thanks, I will drop this

giuseppe replied:
I looked into this, but since I use `self.system` short after, I preferred to keep this as it makes the code cleaner instead of checking if `self.args` has `system` every time.  It is similar to .spc defined just above

rhatdan replied:
No big thing, but I thought we were defaulting self.system=false in the init, so that if this throws the exception, it will still be set to false.  Which is why I suggested to just pass it.

```
          try:
            self.system = args.system
        except:
            pass
```

rhatdan commented on Atomic/util.py:
```diff
@@ -67,22 +67,23 @@ def image_by_name(img_name, images=None):
     if not isinstance(cmd, list):
         cmd = shlex.split(cmd)
-    return subprocess.check_call(cmd, env=env, stderr=stderr, stdout=stdout)
+    return subprocess.check_call(cmd, env=env, stdin=stdin, stderr=stderr, stdout=stdout)
```
We should just merge this whole section separately.

giuseppe replied:
I will split it into another commit

runcom commented:
@giuseppe I still don't like that flag in skopeo cause it doesn't make sense for me for docker images.. Im really not sure, that http: sounds incompatible to me with skopeo and docker images. Skopeo does have flags for tls-verify but it's up to the discovery mechanism to tell if the image is being requested with http or https, @rhatdan wdyt?

rhatdan commented on docs/atomic-update-container.1.md:
```diff
@@ -0,0 +1,34 @@
+% ATOMIC(1) Atomic Man Pages
+% Giuseppe Scrivano
```
As stated above, I think it would be better to expand on the atomic update command

atomic update --container or something like that.

giuseppe replied:
I was in doubt about that, as atomic update accepts an IMAGE as argument, but yes, I will change it to accept the CONTAINER name when --container is used.  It sounds better than another command.  I would still like to make it clear we are updating a system container, is `atomic update --container --system CONTAINER_NAME` fine?

rhatdan replied:
Well as I said on the uninstall, I would prefer it atomic could figure this out for itself.

rhatdan commented:
>@giuseppe I still don't like that flag in skopeo cause it doesn't make sense for me for docker images.. >Im really not sure, that http: sounds incompatible to me with skopeo and docker images. Skopeo >does have flags for tls-verify but it's up to the discovery mechanism to tell if the image is being >requested with http or https, @rhatdan wdyt? 

I would rather http indicates insecure and https indicates secure and not have any special flags.

runcom commented:
Right, but right now skopeo don't have any clue about http vs https, skopeo handles download following the discovery logic of docker images where you don't provide an http or https prefix. Everything is auto discovered pinging the endpoint once with https and the with http. Skopeo (and docker) will always use the discovered scheme when interacting with the registry. I don't want to introduce http or https prefix in skopeo cause it's not how docker does this and since skopeo is meant for docker images it seems awkward to tell people to prefix their images. Skopeo does have flags for tls-verify as said though (same as Docker has)

giuseppe commented:
I have addressed all the comments reported, now it is possible to do:

```
# atomic install --system gscrivano/spc-etcd
# atomic uninstall spc-etcd-system
```

One thing left to discuss:

The `atomic pull` command doesn't checkout the image to disk, but it is committed to the OSTree repository.  `atomic install --system` will do the checkout, and if the image is a Docker imported image, each layer is checked-out, while for a flat OSTree commit it is basically an `ostree checkout` from the commit.  This makes garbage collection easier as OSTree will take care of it, except a part in `atomic images --prune` for removing unused branches, and we won't have to care what containers are using what image.

giuseppe commented:
I fixed a few more things :arrow_up:

rhatdan commented on Atomic/atomic.py:
```diff
@@ -379,6 +379,14 @@ def set_args(self, args):
+        if not self.storage:
+            self.storage = self.get_atomic_config_item(["default_storage"])
+
         if not self.name and self.image is not None:
```
What happens is self.storage is still None?  Will everything still work. or will the system throw a  proper exception?

giuseppe replied:
if storage is null, it will assume `ostree`

rhatdan replied:
Ok great.  Just wanted to make sure you handle this case.

rhatdan commented on atomic.conf:
```diff
@@ -3,3 +3,4 @@
 default_scanner: openscap
 default_docker: docker
 
+# default_storage: ostree
```
Why not make this the default?  What other choices do I have?  Should we list options in the config file?

giuseppe replied:
this is the default when nothing is set as well.  Do you want it to be "default" in the sense of being explicit in the configuration file or also when nothing is set?
At the moment only `ostree` storage is supported.

rhatdan replied:
Fine leave it commented.

rhatdan commented on Atomic/atomic.py:
```diff
@@ -30,6 +41,22 @@
+[Service]
+ExecStart=/bin/runc start $NAME
+Restart=on-failure
+WorkingDirectory=$DESTDIR
```
Should this be Restart=on-crash

On-Failure will continue to attempt to restart if their is a config error, while On-Crash will restart is runc crashes for some reason.

giuseppe replied:
Thanks.  Fixed as fixup patch

rhatdan commented on atomic:
```diff
@@ -277,6 +277,13 @@ if __name__ == '__main__':
+                          action='append',
+                          help=_("Specify a variable in the VARIABLE=VALUE "
+                                 "form for a system container"))
     installp.add_argument("args", nargs=argparse.REMAINDER,
```
If I call --set on a non --system container are we checking for this and exiting out?

rhatdan commented on bash/atomic:
```diff
@@ -94,8 +94,13 @@ __atomic_image_repos_and_tags() {
+		images="$(__docker_q images -a --no-trunc | awk 'NR>1 { print $3; if ($1 != "<none>") { print $1; print $1":"$2 } }')"
+	else
+		images="$(ls -1 /var/lib/containers/atomic/ | grep -v \.[01]$)"
+	fi
```
This path should come from the config, although I would not hold up the merge to get it.

giuseppe replied:
Ouch true, thanks for spotting it. I fixed it in a fixup commit.

rhatdan commented:
@runcom @cgwalters @mrunalp PTAL

We could use some test cases for this also.

giuseppe commented:
fixed :arrow_up: the issues reported.  How do we get to install the missing dependency python3-gobject-base so we can have tests checked on the internal Jenkins?

rhatdan commented:
@giuseppe Looks like you need to refresh your pulls.  Merge some together also, would be helpful.

giuseppe commented:
I rebased the patches :arrow_up:

rhatdan commented:
Looks like we need to add ostree to our build environment.  @cgwalters how do we do that?

@mrunalp @runcom @baude @cgwalters PTAL

mrunalp commented on Atomic/atomic.py:
```diff
@@ -8,6 +8,17 @@
+import json
+import tarfile
+import stat
+import gi
```
This should probably be added to requirements.txt as a new dependency.

mrunalp commented on Atomic/atomic.py:
```diff
@@ -8,6 +8,17 @@
+from string import Template
+gi.require_version('OSTree', '1.0')
+from gi.repository import Gio, GLib, OSTree
+from collections import defaultdict
```
I don't see usage of defaultdict, so probably remove this unused import?

mrunalp commented on Atomic/atomic.py:
```diff
@@ -8,6 +8,17 @@
+import base64
+import shutil
+import tempfile
+import json
```
We have already imported json once above in the file.

mrunalp commented on Atomic/atomic.py:
```diff
@@ -8,6 +8,17 @@
 import getpass
 import argparse
 import subprocess
+import base64
```
unused import

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+    def _system_images(self):
+        repo = self._get_ostree_repo()
+
+        revs = [x for x in repo.list_refs()[1] if x.startswith(OSTREE_OCIIMAGE_PREFIX) and len(x) != len(OSTREE_OCIIMAGE_PREFIX) + 64]
```
Might want to spread this across multiple lines. https://www.python.org/dev/peps/pep-0008/#maximum-line-length

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+        col_out = "{0:%d} {1:64}" % max_column
+        if len(revs) == 0:
+            return
+        self.writeOut(col_out.format("IMAGE", "COMMIT"))
```
We should probably rename writeOut to write_out as it is not pythonic.

giuseppe replied:
adding a new commit to address this

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+
+        def visit(rev):
+            commit = repo.resolve_rev(rev, False)[1]
+            manifest = self._get_commit_metadata (repo, commit, "docker.manifest")
```
whitespace before ( needs to bre removed.

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+            shutil.rmtree(layers_dir)
+        return True
+
+    def _get_commit_metadata (self, repo, rev, key):
```
Consider using @staticmethod annotation for methods that aren't using `self` like this one. I'll try to spot and add comments for other such methods.

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+
+        rev = repo.resolve_rev(imagebranch, False)[1]
+
+        manifest = self._get_commit_metadata (repo, rev, "docker.manifest")
```
Need to remove whitespace before (

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+            return None
+        return metadata[key]
+
+    def _checkout_system_container(self, repo, name, img, deployment, upgrade, values={}):
```
We should probably avoid a mutable default argument here `values={}`
See http://docs.python-guide.org/en/latest/writing/gotchas/

giuseppe replied:
I think this must be explicit as some values must be present, as $NAME and $DESTDIR, which are needed for the systemd unit file.

mrunalp commented on bash/atomic:
```diff
@@ -641,6 +664,7 @@ _atomic() {
 		images
 		migrate
 		mount
+                pull
```
indentation looks off

mrunalp commented on dockertar-sha256-helper.go:
```diff
@@ -0,0 +1,36 @@
+		if err != nil {
+			panic(err)
+		}
+		w.Write(buf[:n])
```
Do we need to remove this line since it got moved up?

giuseppe replied:
ouch, thanks for catching this error.

mrunalp commented on Atomic/atomic.py:
```diff
@@ -261,6 +297,59 @@ def push(self):
 
                 prevstatus = status
 
+
```
don't need this newline

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+                repo.set_ref_immediate(ref[1], ref[2], None)
+        return
+
+    def _parse_imagename(self, imagename):
```
could be static method

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+        else:
+            return reg, image, "latest"
+
+    def _convert_to_skopeo(self, image):
```
could be static method

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+        else:
+            return ["docker://" + i.replace("https:", "")]
+
+    def _skopeo_get_manifest(self, image):
```
could be static method

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+            raise e
+        return temp_dir
+
+    def _get_layers_from_manifest(self, manifest):
```
could be static method

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+            layers = manifest_json.get("Layers")
+        return layers
+
+    def _import_layers_into_ostree(self, repo, imagebranch, manifest, layers):
```
could be static method

mrunalp commented on Atomic/atomic.py:
```diff
@@ -696,6 +820,407 @@ def install(self):
+            finally:
+                shutil.rmtree(temp_dir)
+
+    def _check_system_ostree_image(self, repo, img, upgrade):
```
could be static method

giuseppe commented:
@mrunalp, fixed in this new version :arrow_up:

mrunalp commented:
@giuseppe Thanks! I think that's all the comments I had.

rhatdan commented:
@runcom PTAL

runcom commented:
The only thing I'd like to see here is some integration tests (for such a big and important feature) to not regress in the future, code LGTM otherwise, will provision a vm and manually test this either today evening or tomorrow

rhatdan commented:
I agree tests are critical, but not easy to generate.  If everything looks go to you when you test lets merge and put out atomic-1.10 -rc1  Want to get this out into RHEL  release, which is due tomorrow.

runcom commented:
:+1:

runcom commented:
@giuseppe is there anything missing skopeo side?

giuseppe commented:
@runcom, there is no --insecure yet for oci:http: yet, right?  I think it is fine to keep it as a secret feature for now, and publicize once Skopeo has `--insecure` instead of removing it from the code.  What do you think?

Also, there is no --verify yet, but that is a missing part that we will add later.

runcom commented:
If @rhatdan is fine merging this w/o --verify I'm fine also. I'll add --insecure to skopeo tomorrow

runcom commented:
Will ping you tomorrow morning for other questions

rhatdan commented:
We can add --verify and then just report Not Implemented yet.  Although we need to show something at the Summit.

giuseppe commented:
do you want me to add a patch with that functionality or it can be done later?

giuseppe commented:
ok, I have rebased the patches and added `pull --verify` that passes it directly to Skopeo :arrow_up: 

Does it look fine?

runcom commented:
tested this out, didn't spot anything, LGTM (will work on skopeo now)

giuseppe commented:
@rhatdan what do you think about these last changes?

rhatdan commented on Atomic/atomic.py:
```diff
@@ -390,6 +390,14 @@ def set_args(self, args):
+            self.verify_image = args.verify
+        except:
+            self.verify_image = None
+
```
Should this be false and set by default? Or should we read config file to set the default.

It should definately be a boolean.

rhatdan commented on Atomic/atomic.py:
```diff
@@ -390,6 +390,14 @@ def set_args(self, args):
+
+        if self.verify_image is None:
+            self.verify_image = self.get_atomic_config_item(["verify_image"])
+
```
Should do this when self.verify_image is initialized.

rhatdan commented:
Also what are we verifying?  Perhaps this should take a value.

--verify=redhat.com

Might make sense.

TO verify who you want to sign the image.

rhatdan commented:
Why don't we drop verify for now so we can merge the rest and cut atomic-1.10.  --verify will be in atomic-1.11.

giuseppe commented:
yes, I agree, I removed the commit which added `--verify`

rhatdan commented:
@giuseppe @runcom @mrunalp @baude 
Should we merge this and cut atomic-1.10, and continue doing bug fixes?

giuseppe commented:
@rhatdan I think yes.  Can we also say that it is not yet stable and the interface may break in future?

runcom commented:
that sounds good to me, but I'd love to stick and not break the interface (as I did with skopeo :()

baude commented:
@rhatdan can you give me by EOD to get a PR fix for the changes in skopeo prior to cutting the release?  This will include the v1 inspection handling and the changes in the skopeo's JSON structure.

rhatdan commented:
@baude I will do better then that I will give you to the end of my west coast day.
